### ISSUE TITLE
Google Cloud Storage stream upload

### ISSUE THREAD

**Issue by thdxr** [googleapis/elixir-google-api#968]
I've noticed in my metrics that I see ram spikes whenever we schedule uploads of large files to GCS.  It seems the file is being loaded all into memory first before being uploaded, wanted to confirm if this was the case and if we could enable streaming

**Comment by zeroasterisk** [comment]
Are you using the `storage_objects_insert_simple` method?  
It receives the file content as an arg.

Can you provide me an example use case, how you're using it now, and how you would like to see it, as a stream?

(I agree with this request... but I'll have to see what streaming/chunking options I can wire in)

**Comment by thdxr** [comment]
I'm currently passing a file path pointing to the file to `storage_objects_insert_simple` 

Can I pass a stream instead>

**Comment by zeroasterisk** [comment]
No. It would still consume the entire file into a string to send as a POST
to the Google API.

I would love to research this and determine how to either stream over a
socket, or multi-part chunk the file... But I'm not sure when I'll get to
it.

Is this a blocker for you?

**Comment by thdxr** [comment]
Not a blocker yet. File sizes aren't unreasonably big yet but probably will become an issue in a few months. No rush

**Comment by col** [comment]
:+1:

**Comment by mathiasose** [comment]
I've run into this problem now, the data I want to upload as a single file is too large for memory. It's actually coming out of the DB as a lazy stream, but as mentioned above it seems the upload function (I'm using `storage_objects_insert_iodata`) is consuming the whole stream into memory before doing any network transfers.

Are there APIs for streaming/chunked uploads on the gcloud side? Is it just a matter of implementing an interface for these if they exist?

**Comment by spencerkent** [comment]
I'd really benefit from this too. It seems like it's [already supported](https://cloud.google.com/storage/docs/samples/storage-stream-file-upload#storage_stream_file_upload-python) in APIs for other languages 🤔

**Comment by rozap** [comment]
It also looks like even if you do have quite a lot of memory, `storage_objects_insert_iodata` just returns `{:error, :closed}` for large files.
